disco_aws_automation/disco_elasticsearch_archive.py

```diff
@@ -0,0 +1,498 @@
+
+    def _archivable_indices(self):
+        """ Return list of indices that are older than today's date """
+        # TODO: might want to consider timezone issue?
+        today = datetime.date.today().strftime('%Y.%m.%d')
+
+        return [index[0]
+                for index in self._get_all_indices_and_sizes()
+                if index[0][-10:] < today]
+
+    def _get_all_indices_and_sizes(self):
+        """
+        Return list of all the indices and their sizes sorted by the date from
+        oldest to latest
+        """
+        stats = self.es_client.indices.stats()
+        dated_index_pattern = re.compile(
+            self.index_prefix_pattern + r'-[\d]{4}(\.[\d]{2}){2}$'
+        )
+        index_sizes = [
+            (index, stats['indices'][index]['total']['store']['size_in_bytes'])
+            for index in sorted(stats['indices'], key=lambda k: k[-10:])
+            if re.match(dated_index_pattern, index)
+        ]
+
+        return index_sizes
+
+    def _green_indices(self):
+        """
+        Return list of indexes that are in green state (healthy)
+        """
+        index_health = self.es_client.cluster.health(level='indices')
+        return [
+            index
+            for index in index_health['indices'].keys()
+            if index_health['indices'][index]['status'] == 'green'
+        ]
+
+    def _create_repository(self):
+        """
+        Creates a s3 type repository where archives can be stored
+        and retrieved.
+        """
+
+        # Make sure S3 bucket is available by updating it.
+        self._update_s3_bucket(self.s3_bucket_name)
+
+        repository_config = {
+            "type": self.repository_name,
+            "settings": {
+                "bucket": self.s3_bucket_name,
+                "region": self.region,
+                "role_arn": self.role_arn,
+            }
+        }
+
+        logging.info("Creating ES snapshot repository: %s", self.repository_name)
+        try:
+            self.es_client.snapshot.create_repository(self.repository_name, repository_config)
+        except TransportError:
+            try:
+                self.es_client.snapshot.get_repository(self.repository_name)
+                # This can happen when repo is used to make snapshot but we
+                # attempt to re-create it.
+                logging.warning(
+                    "Error on creating ES snapshot respository %s, "
+                    "but one already exists, ingnoring",
+                    self.repository_name
+                )
+            except:
+                logging.error(
+                    "Could not create archive repository. "
+                    "Make sure bucket %s exists and IAM policy allows es to access bucket. "
+                    "repository config: %s",
+                    bucket_name,
+                    repository_config,
+                )
+                raise
+
+    def _update_s3_bucket(self, bucket_name):
```

Comment:
I don't think we have disco_s3. Currently, we are using the shell script below to create S3 buckets:

https://github.com/amplifylitco/asiaq/blob/master/jenkins/disco_bucket_update.sh

I am actually wondering if we should also use that to create the S3 buckets for ES archival, for consistency's sake and avoiding code redundancy. If the S3 bucket is not available ES archival is run, we should just error out. Thoughts?